src/language_features/document_links.rs

```diff
@@ -0,0 +1,213 @@
+    pub selection_desc: String,
+}
+
+pub fn resolve_and_follow_document_link(
+    meta: EditorMeta,
+    params: DocumentLinkOptions,
+    ctx: &mut Context,
+) {
+    let (range, _cursor) = parse_kakoune_range(&params.selection_desc);
+    let document = match ctx.documents.get(&meta.buffile) {
+        Some(document) => document,
+        None => return,
+    };
+
+    if let Some((server_id, link)) = ctx
+        .document_links
+        .get(&meta.buffile)
+        .and_then(|links| {
+            links.iter().find(|(server_id, link)| {
+                let ServerSettings {
+                    offset_encoding, ..
+                } = ctx.server(*server_id);
+                let range = kakoune_range_to_lsp(&range, &document.text, *offset_encoding);
+                ranges_overlap(link.range, range)
+            })
+        })
+        .filter(|(_, link)| link.target.is_none())
+        .cloned()
+    {
+        let mut req_params = HashMap::new();
+        req_params.insert(server_id, vec![link]);
+
+        ctx.call::<DocumentLinkResolve, _>(
+            meta,
+            RequestParams::Each(req_params),
+            |ctx: &mut Context, meta, results| follow_document_link(meta, &results, ctx),
+        );
+        return;
+    }
+
+    let no_links = vec![];
+    let links = ctx.document_links.get(&meta.buffile).unwrap_or(&no_links);
+    let mut links = links
+        .iter()
+        .filter(|(server_id, link)| {
+            let ServerSettings {
+                offset_encoding, ..
+            } = ctx.server(*server_id);
+            let range = kakoune_range_to_lsp(&range, &document.text, *offset_encoding);
+            ranges_overlap(link.range, range)
+        })
+        .map(|(a, b)| (*a, b.clone()))
+        .collect::<Vec<_>>();
+
+    links.sort_by_key(|(_server_name, link)| {
+        let Range { start, end } = link.range;
+        end.line - start.line
+    });
+
+    follow_document_link(meta, &links, ctx);
+}
+
+fn follow_document_link(meta: EditorMeta, links: &[(ServerId, DocumentLink)], ctx: &mut Context) {
+    let commands = links
+        .iter()
+        .filter(|(_, link)| link.target.is_some())
+        .map(|(_, link)| {
+            let target = link.target.as_ref().unwrap();
+            let label = editor_quote(target.as_str());
+            let path = Some(target)
+                .filter(|x| x.scheme() == "file")
+                .and_then(|x| x.to_file_path().ok())
+                .map(|x| x.to_string_lossy().to_string());
+            let command = match path {
+                Some(path) => format!("edit -existing {}", &editor_quote(&path)),
+                None => format!("lsp-open-url {}", &editor_quote(target.as_str())),
+            };
+            (label, command)
+        })
+        .collect_vec();
+
+    match commands.len() {
+        0 => ctx.show_error(meta, "no document link in selection"),
+        1 => ctx.exec(meta, commands.into_iter().next().unwrap().1),
```

Comment:
(I think we do support `menu -auto-single` but I agree with explicitly short-circuiting here)